app/Models/ProductModel.php

```diff
@@ -2,17 +2,107 @@
+     */
     public function getAllProducts()
     {
         return $this->select('productId, productName, productModel, categoryName, fileRoute')->join('product_files', 'product_files.pfProductId = products.productId')->join('files', 'files.fileId = product_files.pfFileId')->join('categories', 'categories.categoryId = products.productCategoryId')->where('pfFileType', 'main image')->findAll();
     }
+
+    /**
+     * Create the query for filter products by categories, categoryTags, productTags and a search by productName or productModel
+     *
+     * @param array  $categories   all categories to filter
+     * @param array  $categoryTags all categoryTags to filter
+     * @param array  $productTags  all productTags to filter
+     * @param string $search       the product name or product model to filter
+     *
+     * @return $this
+     */
+    public function filterProducts(array $categories, array $categoryTags, array $productTags, string $search = '')
+    {
+        $builder = $this->select([
+            'productName',
+            'productModel',
+            'categoryName',
+            'CONCAT("/producto/", productId, "/", productName) AS productUrl',
+            'GROUP_CONCAT(DISTINCT files.fileRoute) AS imageRoute',
+        ])
+            ->join('categories', 'products.productCategoryId = categories.categoryId')
+            ->join('`products-category_tags`', 'products.productId = `products-category_tags`.pctProductId', 'LEFT')
+            ->join('category_tags', 'category_tags.categoryTagId = `products-category_tags`.pctCategoryTagId', 'LEFT')
+            ->join('product_tags', 'product_tags.ptProductId = products.productId', 'LEFT')
+            ->join('product_files', new RawSql('products.productId = product_files.pfProductId AND product_files.pfFileType = "main image"'), 'LEFT')
+            ->join('files', 'product_files.pfFileId = files.fileId', 'LEFT');
+        $whereQueries = [];
+
+        // added categories filter
+        if (! empty($categories)) {
+            $whereQueries[] = "categories.categorySlug IN ('" . str_replace(' ', '', implode("', '", $categories)) . "')";
+        }
+        // added category_tags filter
+        if (! empty($categoryTags)) {
+            $whereQueries[] = "category_tags.categoryTagSlug IN ('" . str_replace(' ', '', implode("', '", $categoryTags)) . "')";
+        }
+        // added products_tags filter
+        if (! empty($productTags)) {
+            $whereQueries[] = "product_tags.ptSlug IN ('" . str_replace(' ', '', implode("', '", $productTags)) . "')";
+        }
+        // added search filter
+        if (! empty($search)) {
+            $whereQueries[] = "(products.productName LIKE '%{$search}%' OR products.productModel LIKE '%{$search}%')";
+        }
+
+        $builder->groupBy('productId');
+
+        if (empty($whereQueries)) {
+            return $this;
+        }
+        $whereQuery = implode(' AND ', $whereQueries);
+        $builder->where($whereQuery);
```

Comment:
Instead of creating `where in` statements manually, you can make use of the whereIn method provided by the CI query builder, and the same case with the like statement.

```suggestion
             // added categories filter
        if (! empty($categories)) {
            $builder->whereIn('categories.categorySlug', $categories);
        }
        // added category_tags filter
        if (! empty($categoryTags)) {
            $builder->whereIn('category_tags.categoryTagSlug', $categoryTags);
        }
        // added products_tags filter
        if (! empty($productTags)) {
            $builder->whereIn('product_tags.ptSlug', $productTags);
        }
        // added search filter
        if (! empty($search)) {
            $builder->like('products.productName', $search)->orLike('products.productModel', $search);
        }

        $builder->groupBy('productId');

        return $this;
```